### ISSUE TITLE
Conv2d with DRAM Slice Config Drops PCC for `act_block_h_override=0`

### ISSUE THREAD

**Issue by bmalesevicTT** [tenstorrent/tt-metal#35852]
## Summary
When using `Conv2dSliceConfig` with DRAM slice types (DRAMSliceHeight or DRAMSliceWidth) and the default `act_block_h_override=0`, conv2d operations execute without errors but produce incorrect results with ~50% PCC degradation. This only occurs when using `prepare_conv_weights` and `prepare_conv_bias`. The issue is particularly problematic because it fails silently, unlike L1 slice config which throws a runtime error that the optimizer can catch and handle.

## Background
The default slice config is set to L1, but when operations cannot fit in L1 memory, we need to fallback to DRAM slice configs. However, this fallback currently produces incorrect results silently, making it unusable for automatic fallback in the optimizer.

## Expected Behavior
When `act_block_h_override=0` is incompatible with DRAM slice config, the operation should either:
1. **Raise a runtime error** - Similar to L1 slice config which throws an L1 memory overflow error, allowing the optimizer to catch it and adjust `act_block_h_override` automatically
2. **Execute correctly** - Handle the default value properly and produce correct results

## Actual Behavior
The operation executes without errors but produces incorrect results with PCC dropping from &gt;0.99 to ~0.5.

## Steps to Reproduce

```python
import torch
import ttnn

# Conv2d parameters
batch_size = 1
in_channels = 3
out_channels = 32
input_height = 480
input_width = 640
kernel_size = [3, 3]
stride = [1, 1]
padding = [1, 1, 1, 1]

device = ttnn.open_device(device_id=0)

try:
    # Create tensors
    input_torch = torch.randn((batch_size, in_channels, input_height, input_width), dtype=torch.bfloat16)
    weight_torch = torch.randn((out_channels, in_channels, kernel_size[0], kernel_size[1]), dtype=torch.bfloat16)
    bias_torch = torch.randn((out_channels,), dtype=torch.bfloat16)

    # Reshape to conv2d format
    input_torch_nhwc = input_torch.permute(0, 2, 3, 1).contiguous()
    input_torch_reshaped = input_torch_nhwc.reshape(batch_size, 1, input_height * input_width, in_channels)
    bias_torch_reshaped = bias_torch.reshape(1, 1, 1, out_channels)

    # Convert to TTNN tensors
    input_ttnn = ttnn.from_torch(input_torch_reshaped, dtype=ttnn.bfloat16, 
                                  layout=ttnn.TILE_LAYOUT, device=device, 
                                  memory_config=ttnn.DRAM_MEMORY_CONFIG)
    weight_ttnn = ttnn.from_torch(weight_torch, dtype=ttnn.bfloat16, 
                                   layout=ttnn.ROW_MAJOR_LAYOUT, device=None)
    bias_ttnn = ttnn.from_torch(bias_torch_reshaped, dtype=ttnn.bfloat16, 
                                 layout=ttnn.ROW_MAJOR_LAYOUT, device=None)

    # PROBLEMATIC CONFIG: DRAM slice with act_block_h_override=0
    conv2d_config = ttnn.Conv2dConfig(
        weights_dtype=ttnn.bfloat16,
        deallocate_activation=True,
        act_block_h_override=0,  # Default value - causes silent failure with DRAM slice
        enable_kernel_stride_folding=False,
        config_tensors_in_dram=True
    )

    conv2d_slice_config = ttnn.Conv2dSliceConfig(
        slice_type=ttnn.Conv2dSliceConfig.SliceTypeEnum.DRAMSliceHeight,  # Also fails with DRAMSliceWidth
        num_slices=0
    )

    # Prepare weights and bias (required to trigger the issue)
    prepared_weight = ttnn.prepare_conv_weights(
        weight_tensor=weight_ttnn,
        input_memory_config=ttnn.DRAM_MEMORY_CONFIG,
        input_layout=ttnn.TILE_LAYOUT,
        weights_format="OIHW",
        in_channels=in_channels,
        out_channels=out_channels,
        batch_size=batch_size,
        input_height=input_height,
        input_width=input_width,
        kernel_size=kernel_size,
        stride=stride,
        padding=padding,
        dilation=[1, 1],
        has_bias=True,
        groups=1,
        device=device,
        input_dtype=ttnn.bfloat16,
        output_dtype=ttnn.bfloat16,
        conv_config=conv2d_config,
        slice_config=conv2d_slice_config
    )

    prepared_bias = ttnn.prepare_conv_bias(
        bias_tensor=bias_ttnn,
        input_memory_config=ttnn.DRAM_MEMORY_CONFIG,
        input_layout=ttnn.TILE_LAYOUT,
        in_channels=in_channels,
        out_channels=out_channels,
        batch_size=batch_size,
        input_height=input_height,
        input_width=input_width,
        kernel_size=kernel_size,
        stride=stride,
        padding=padding,
        dilation=[1, 1],
        device=device,
        input_dtype=ttnn.bfloat16,
        output_dtype=ttnn.bfloat16,
        groups=1,
        conv_config=conv2d_config,
        slice_config=conv2d_slice_config
    )

    # Run conv2d - executes without error but produces wrong results
    output_ttnn = ttnn.conv2d(
        input_tensor=input_ttnn,
        weight_tensor=prepared_weight,
        bias_tensor=prepared_bias,
        device=device,
        in_channels=in_channels,
        out_channels=out_channels,
        batch_size=batch_size,
        input_height=input_height,
        input_width=input_width,
        kernel_size=kernel_size,
        stride=stride,
        padding=padding,
        dilation=[1, 1],
        groups=1,
        conv_config=conv2d_config,
        slice_config=conv2d_slice_config,
        memory_config=ttnn.DRAM_MEMORY_CONFIG,
        dtype=ttnn.bfloat16
    )

    # Reshape output back [1, 1, 307200, 32] -&gt; [1, 480, 640, 32]
    output_ttnn_reshaped = ttnn.reshape(output_ttnn, (batch_size, input_height, input_width, out_channels))

    # Permute to NCHW [1, 480, 640, 32] -&gt; [1, 32, 480, 640]
    output_ttnn_nchw = ttnn.permute(output_ttnn_reshaped, (0, 3, 1, 2))

    # Convert back to torch
    output_torch = ttnn.to_torch(output_ttnn_nchw)

    # Compute reference output using PyTorch
    with torch.no_grad():
        ref_output = torch.nn.functional.conv2d(
            input_torch,
            weight_torch,
            bias_torch,
            stride=stride[:2],
            padding=(padding[0], padding[2]),
            dilation=[1, 1],
            groups=1
        )

    # Compute PCC
    def compute_pcc(tensor1, tensor2):
        tensor1_flat = tensor1.flatten().float()
        tensor2_flat = tensor2.flatten().float()
        mean1 = tensor1_flat.mean()
        mean2 = tensor2_flat.mean()
        std1 = tensor1_flat.std()
        std2 = tensor2_flat.std()
        covariance = ((tensor1_flat - mean1) * (tensor2_flat - mean2)).mean()
        pcc = covariance / (std1 * std2 + 1e-8)
        return pcc.item()

    pcc = compute_pcc(output_torch.cpu(), ref_output.cpu())
    print(f"PCC: {pcc:.6f}")
    
    if pcc &gt; 0.99:
        print("✓ Test PASSED - Good PCC")
    else:
        print("✗ Test FAILED - Bad PCC")  # This will be printed with act_block_h_override=0 + DRAM slice

finally:
    ttnn.close_device(device)
```
</code></pre></div><p style="margin-top: 0.1em; margin-bottom: 0.2em; white-space: pre-wrap; color: rgb(204, 204, 204); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><strong>Expected Output with act_block_h_override=1024 (working):</strong></p><div class="Q" style="position: relative; margin: 8px 0px; color: rgb(204, 204, 204); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><button class="J Co" title="Copy code" aria-label="Copy code to clipboard" style="color: rgb(204, 204, 204); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; padding: 4px; background: none 0% 0% / auto repeat scroll padding-box border-box rgb(31, 31, 31); border-color: rgb(69, 69, 69); border-style: solid; border-width: 1px; border-image: none 100% / 1 / 0 stretch; border-radius: 4px; cursor: pointer; opacity: 0; transition: opacity 0.15s, background 0.15s; display: flex; align-items: center; justify-content: center; position: absolute; top: 4px; right: 4px;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" aria-hidden="true" data-slot="icon" class="Kr"><path fill-rule="evenodd" d="M15.988 3.012A2.25 2.25 0 0 1 18 5.25v6.5A2.25 2.25 0 0 1 15.75 14H13.5v-3.379a3 3 0 0 0-.879-2.121l-3.12-3.121a3 3 0 0 0-1.402-.791 2.252 2.252 0 0 1 1.913-1.576A2.25 2.25 0 0 1 12.25 1h1.5a2.25 2.25 0 0 1 2.238 2.012ZM11.5 3.25a.75.75 0 0 1 .75-.75h1.5a.75.75 0 0 1 .75.75v.25h-3v-.25Z" clip-rule="evenodd"></path><path d="M3.5 6A1.5 1.5 0 0 0 2 7.5v9A1.5 1.5 0 0 0 3.5 18h7a1.5 1.5 0 0 0 1.5-1.5v-5.879a1.5 1.5 0 0 0-.44-1.06L8.44 6.439A1.5 1.5 0 0 0 7.378 6H3.5Z"></path></svg></button><pre style="overflow-x: auto; max-width: 100%; white-space: pre; padding: 8px; border-radius: 4px; margin: 0px; box-sizing: border-box;"><code style="font-family: monospace; color: rgb(208, 208, 208); background-color: rgb(60, 60, 60); padding: 0px; border-radius: 3px; font-size: 0.9em; word-break: break-word;">PCC: 0.999xxx
✓ Test PASSED - Good PCC
</code></pre></div><p style="margin-top: 0.1em; margin-bottom: 0.2em; white-space: pre-wrap; color: rgb(204, 204, 204); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><strong>Actual Output with act_block_h_override=0 (silent failure):</strong></p><div class="Q" style="position: relative; margin: 8px 0px; color: rgb(204, 204, 204); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><button class="J Co" title="Copy code" aria-label="Copy code to clipboard" style="color: rgb(204, 204, 204); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, sans-serif; font-size: 13px; padding: 4px; background: none 0% 0% / auto repeat scroll padding-box border-box rgb(31, 31, 31); border-color: rgb(69, 69, 69); border-style: solid; border-width: 1px; border-image: none 100% / 1 / 0 stretch; border-radius: 4px; cursor: pointer; opacity: 0; transition: opacity 0.15s, background 0.15s; display: flex; align-items: center; justify-content: center; position: absolute; top: 4px; right: 4px;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" aria-hidden="true" data-slot="icon" class="Kr"><path fill-rule="evenodd" d="M15.988 3.012A2.25 2.25 0 0 1 18 5.25v6.5A2.25 2.25 0 0 1 15.75 14H13.5v-3.379a3 3 0 0 0-.879-2.121l-3.12-3.121a3 3 0 0 0-1.402-.791 2.252 2.252 0 0 1 1.913-1.576A2.25 2.25 0 0 1 12.25 1h1.5a2.25 2.25 0 0 1 2.238 2.012ZM11.5 3.25a.75.75 0 0 1 .75-.75h1.5a.75.75 0 0 1 .75.75v.25h-3v-.25Z" clip-rule="evenodd"></path><path d="M3.5 6A1.5 1.5 0 0 0 2 7.5v9A1.5 1.5 0 0 0 3.5 18h7a1.5 1.5 0 0 0 1.5-1.5v-5.879a1.5 1.5 0 0 0-.44-1.06L8.44 6.439A1.5 1.5 0 0 0 7.378 6H3.5Z"></path></svg></button><pre style="overflow-x: auto; max-width: 100%; white-space: pre; padding: 8px; border-radius: 4px; margin: 0px; box-sizing: border-box;"><code style="font-family: monospace; color: rgb(208, 208, 208); background-color: rgb(60, 60, 60); padding: 0px; border-radius: 3px; font-size: 0.9em; word-break: break-word;">PCC: 0.5xxxxx
✗ Test FAILED - Bad PCC
</code></pre></div><h2 style="color: rgb(204, 204, 204); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(24, 24, 24); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Comparison of Behaviors</h2>
Configuration | act_block_h_override=0 | act_block_h_override=1024
-- | -- | --
L1 slice config | ❌ Runtime error: Statically allocated circular buffers on core range [(x=0,y=0) - (x=7,y=6)] grow to 2235456 B which is beyond max L1 size of 1499136 B | ✅ Works correctly
DRAM slice config | ⚠️ Silently fails (bad PCC ~0.5) | ✅ Works correctly
DRAM slice without prepare_weights/bias | ✅ Works correctly | ✅ Works correctly

## Workaround
Set `act_block_h_override=1024` explicitly when using DRAM slice configs with prepared weights/bias.

## Impact
This issue prevents automatic fallback from L1 to DRAM slice configs in the optimizer. With L1 slice config, the optimizer can catch the runtime error about L1 memory overflow and adjust `act_block_h_override` accordingly. However, with DRAM slice config, there is no error to catch - the operation silently produces incorrect results instead.

## Environment
- Device: Wormhole
- Input shape: [1, 3, 480, 640]
- Conv2d: 32 output channels, 3x3 kernel

## TLDR
- The issue occurs with both `DRAMSliceHeight` and `DRAMSliceWidth` slice types
- Without using `prepare_conv_weights` and `prepare_conv_bias`, the operation works correctly with either value of `act_block_h_override`
- The L1 slice type config throws a runtime error when circular buffers exceed L1 memory limits with `act_block_h_override=0`, which allows the optimizer to detect the issue and retry with a different value
- DRAM slice config should exhibit similar behavior - either throw an error or execute correctly, but not silently produce wrong results

**Comment by s-jovic** [comment]
Hey @sankarmanoj-tt, do you have some insights on this issue?

**Comment by sankarmanoj-tt** [comment]
Failures with prepare weights are typically due to a mismatch between the configuration chosen by the auto slice + auto shard logic of prepare weights, vs the actual conv2d function. 
The best way to confirm is by running this with the Logger Level set to trace, and looking at what configuration each function selects. 
The most likely underlying cause is because of a mismatch in arguments to each function. 

This PR https://github.com/tenstorrent/tt-metal/pull/35163 solves this issue by combining the prepared weights with it's associated configuration.

**Comment by bmalesevicTT** [comment]
@sankarmanoj-tt Is there any progress on PR #35163 that addresses the prepare weights configuration mismatch issue? 

Aside for this case, where I have a workaround, I suspect issue #36487 may also be affected by the same underlying prepare weights bug, and for that case I haven't found a viable workaround, except fully avoiding prepare weights - which causes significant performance degradation, making it impractical for our use-case.

Would appreciate any update on the timeline for this fix, or if there are alternative approaches to consider in the meantime.

**Comment by s-jovic** [comment]
@sankarmanoj-tt moved to another team - he is not working on the convolutions anymore.

As Sankar said, the problem is that we choose different configurations when preparing weights vs when actually calling conv2d in runtime. This happens because when we are preparing weights we have no information about the memory usage at the time of conv2d op execution. To avoid this problem you would either need to stop using prepare weights, or have some sort of allocator tracking memory usage.

Adding @pilkicTT, @odjuricicTT and @pavlejosipovic as they already had some discussions about this isse.
